model/task/dependency_graph.go

```diff
@@ -0,0 +1,280 @@
+package task
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/multi"
+	"gonum.org/v1/gonum/graph/topo"
+	"gonum.org/v1/gonum/graph/traverse"
+)
+
+// DependencyGraph models task dependency relationships as a directed graph.
+// Use NewDependencyGraph() to initialize a new DependencyGraph.
+type DependencyGraph struct {
+	graph               *multi.DirectedGraph
+	tasksToNodes        map[TaskNode]graph.Node
+	nodesToTasks        map[graph.Node]TaskNode
+	edgesToDependencies map[edgeKey]DependencyEdge
+}
+
+// NewDependencyGraph returns an initialized DependencyGraph.
+func NewDependencyGraph() DependencyGraph {
+	return DependencyGraph{
+		graph:               multi.NewDirectedGraph(),
+		tasksToNodes:        make(map[TaskNode]graph.Node),
+		nodesToTasks:        make(map[graph.Node]TaskNode),
+		edgesToDependencies: make(map[edgeKey]DependencyEdge),
+	}
+}
+
+type edgeKey struct {
+	from TaskNode
+	to   TaskNode
+}
+
+// DependencyEdge is a representation of a dependency in the graph.
+type DependencyEdge struct {
+	// Status is the status specified by the dependency, if any.
+	Status string
+	// From is the node the edge begins from.
+	From TaskNode
+	// To is the node the edge points to.
+	To TaskNode
+}
+
+// TaskNode is the representation of a task in the graph.
+type TaskNode struct {
+	// Name is the display name of the task.
+	Name string
+	// Variant is the build variant of the task.
+	Variant string
+	// ID is the task's ID.
+	ID string
+}
+
+// String represents TaskNode as a string.
+func (t TaskNode) String() string {
+	if t.ID != "" {
+		return t.ID
+	}
+
+	return fmt.Sprintf("%s/%s", t.Variant, t.Name)
+}
+
+// VersionDependencyGraph finds all the tasks from the version given by versionID and constructs a DependencyGraph from them.
+func VersionDependencyGraph(versionID string) (DependencyGraph, error) {
+	tasks, err := FindAllTasksFromVersionWithDependencies(versionID)
+	if err != nil {
+		return DependencyGraph{}, errors.Wrapf(err, "getting tasks for version '%s'", versionID)
+	}
+
+	return taskDependencyGraph(tasks), nil
+}
+
+func taskDependencyGraph(tasks []Task) DependencyGraph {
+	g := NewDependencyGraph()
+	g.buildFromTasks(tasks, false)
+	return g
+}
+
+// reversedTaskDependencyGraph constructs a DependencyGraph from tasks.
+// Edges in the graph are reversed so they point from depended on tasks to dependent tasks.
```

Comment:
It will come in handy in future PRs. 

I hope to use it in the dispatcher which uses a topological sort to put depended on tasks before the tasks that depend on them. Since a topological sort orders the nodes such that for every edge `u -> v ` u comes before v we want u to be the depended on tasks and v to be the dependent tasks. I puzzled over whether it would be equivalent to use the regular graph and just reverse the result. I found [this](https://www.quora.com/Is-a-topological-order-on-the-reverse-graph-the-same-as-a-topological-order-reversed) which proves (?) that it is (and reminded me not to attempt an MS in CS), but it would be an additional effort to get it to work for the case where the topological ordering is ambiguous and we want it to default to the queue order.

It's also applicable to [getRecursiveDependenciesDown](https://github.com/evergreen-ci/evergreen/blob/31a6c66aca73f769b8b003a89c4bffaa529b3b0a/model/task/task.go#L2110-L2113). With a reversed graph we can just call `reachableFromNode` to find the recursive dependencies.

Reply:
I understood just enough of that to let this go :joy: